### PR TITLE
Fix: unable to deserialize a link file to LinkMetadata

### PR DESCRIPTION
**Pull request by Xynnn007**
For https://github.com/in-toto/in-toto-rs/issues/30
Together, also fix when `environment` field in linkfile is null in json, the LinkMetadata can not be deserialized correctly.

Before, the following can not pass `serde_json::deserialize`
```json
{
    "signed": {
        "_type": "link",
        "name": "",
        "materials": {},
        "products": {
            "foo.tar.gz": {
                "sha256": "52947cb78b91ad01fe81cd6aef42d1f6817e92b9e6936c1e5aabb7c98514f355"
            }
        },
        "byproducts": {
            "return-value": 0,
            "stderr": "a foo.py\n",
            "stdout": ""
        },
        "command": [
            "tar",
            "zcvf",
            "foo.tar.gz",
            "foo.py"
        ],
        "environment": null
    },
    "signatures": null
}
```

### CODE REVIEW

File: src/models/link/byproducts.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    /// Set stdout
+    pub fn set_stdout(mut self, stdout: String) -> Self {
+        self.stdout = stdout;
+        self
+    }
+
+    /// Set byproducts
+    pub fn set_other_set_byproducts(mut self, byproducts: BTreeMap<String, String>) -> Self {
```
Comment (cutecutecat) [inline src/models/link/byproducts.rs:59]: also here, might be `set_dynamic_field` or `set_other_field`?

**Comment by cutecutecat** [comment]
You are right, it is indeed a big historical problem need to be fixed.
I agree with it, maybe there will also some similar work about the key `environment` for us to do in the future.

File: src/models/link/byproducts.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    _byproducts: BTreeMap<String, String>,
+}
+
+impl ByProducts {
+    pub fn new() -> Self {
+        ByProducts {
+            return_value: 0,
+            stderr: "".into(),
+            stdout: "".into(),
+            _byproducts: BTreeMap::new(),
```
[the remark this answers is not in the document]
Comment (Xynnn007) [reply src/models/link/byproducts.rs]: Nice suggestion, I've changed to `other_fields`

File: src/models/link/byproducts.rs (hunk 1 above)
Comment (Xynnn007) [reply src/models/link/byproducts.rs]: As mentioned, `set_other_field()` for setting key-value, and `set_other_fields()` for a whole BTreeMap

**Review by cutecutecat — APPROVED** [review]
LGTM.

Though there might be some compatibility problem with previous version, as key `env` is changed to `environment`, I 'm not sure whether this will cause some problem. Apart from this, other change should be compatible in my inference.

I am not familiar about the practical application scenarios. Let's see whether this problem matters by other reviews' view.

File: src/models/link/byproducts.rs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        }"#;
+
+        let byproducts = ByProducts::new()
+            .set_return_value(0)
+            .set_stderr("a foo.py\n".into())
+            .set_stdout("".into());
+
+        let deserialized_byproducts: ByProducts = serde_json::from_str(json).unwrap();
+        assert_eq!(byproducts, deserialized_byproducts);
+    }
```
Comment (SantiagoTorres) [inline src/models/link/byproducts.rs:128]: could we add a test with one extra field as well?

**Review by SantiagoTorres — APPROVED** [review]
I could do with or without that test tbh

**Review by adityasaky — APPROVED** [review]

**Pull request merged**